### PR TITLE
fix: Add reachable classes to child route

### PR DESCRIPTION
**Pull request by mshabarov**
## Description

When child entry-point has no AtRoute annotation, but it's parent class does have, the reachable classes from parent class are added to the child entry-point, so frontend modules are added as well and bundle chunk is generated with all needed web-components.

Fixes https://github.com/vaadin/flow/issues/17722

## Type of change

- [x] Bugfix
- [ ] Feature

## Checklist

- [x] I have read the contribution guide: https://vaadin.com/docs/latest/guide/contributing/overview/
- [x] I have added a description following the guideline.
- [x] The issue is created in the corresponding repository and I have referenced it.
- [x] I have added tests to ensure my change is effective and works as intended.
- [x] New and existing tests are passing locally with my change.
- [x] I have performed self-review and corrected misspellings.

#### Additional for `Feature` type of change

- [ ] Enhancement / new feature was discussed in a corresponding GitHub issue and Acceptance Criteria were created.

### CODE REVIEW

File: flow-server/src/main/java/com/vaadin/flow/server/frontend/scanner/FrontendDependencies.java [hunk 1]
```diff
@@ -216,8 +216,11 @@ private void collectReachableClasses(String name, Set<String> classes) {
 
         classes.add(name);
         for (String className : visitedClass.children) {
-            if (!entryPoints.containsKey(className)) {
+            EntryPointData entryPointData = entryPoints.get(className);
+            if (entryPointData == null) {
                 collectReachableClasses(className, classes);
+            } else if (entryPointData.reachableClasses != null) {
+                classes.addAll(entryPointData.reachableClasses);
             }
```
Comment (mcollovati) [inline flow-server/src/main/java/com/vaadin/flow/server/frontend/scanner/FrontendDependencies.java:224]: I think that there's an uncovered case here: the above code works if the parent class is visited before the child one (`entryPointData.reachableClasses` will not be null), but will not work correctly if the child class is visited before the parent.
And this could happen because both classes are entry points, but I think there's no guarantee about the order they are fetched from the dependency scanner.

We could probably do some sort based on hierarchy in `collectEntryPoints()` after call to `getFinder().getAnnotatedClasses(routeClass)`.

```
        for (Class<?> route : getFinder().getAnnotatedClasses(routeClass)) {
            List<String> triggerClasses = getDependencyTriggers(route,
                    triggerClass);
            boolean eager = isEagerRoute(route);
            addEntryPoint(route, EntryPointType.ROUTE, triggerClasses, eager);
        }
```

Comment (mcollovati) [reply flow-server/src/main/java/com/vaadin/flow/server/frontend/scanner/FrontendDependencies.java:224]: An idea could be to sort in a way that annotated classes gets first (`Class.getDeclaredAnnotation(Route.class) != null)`)

Comment (mshabarov) [reply flow-server/src/main/java/com/vaadin/flow/server/frontend/scanner/FrontendDependencies.java:224]: Excellent catch and good suggestion! Those with annotation go to beginning of the collection, others go to the end.

Comment (mshabarov) [reply flow-server/src/main/java/com/vaadin/flow/server/frontend/scanner/FrontendDependencies.java:224]: And this is most likely the reason why my new test fails.

File: flow-server/src/test/java/com/vaadin/flow/server/frontend/scanner/FrontendDependenciesTest.java [hunk 2]
```diff
@@ -323,6 +324,35 @@ public void shouldVisit_shouldNotMatchOnPartOfPackage() {
                 dependencies.shouldVisit("com.sun"));
     }
 
+    @Test
+    public void classScanningForChildAndParentEntryPoint_parentAnnotated_childNotAnnotated_childHasModulesFromParent() {
+        Mockito.when(classFinder.getAnnotatedClasses(Route.class))
+                .thenReturn(Stream.of(ChildRoute.class, ParentRoute.class)
+                        .collect(Collectors.toSet()));
```
Comment (mcollovati) [inline flow-server/src/test/java/com/vaadin/flow/server/frontend/scanner/FrontendDependenciesTest.java:331]: It would be nice to collect into an `LinkedHashMap` and have two tests, with route classes in different order

File: flow-server/src/main/java/com/vaadin/flow/server/frontend/scanner/FrontendDependencies.java [hunk 3]
```diff
@@ -869,4 +877,18 @@ private boolean hasThemeInfo(ClassInfo classInfo) {
...
+                .getDeclaredAnnotation(routeClass);
+        Annotation routeAnnotation2 = entryPoint2
+                .getDeclaredAnnotation(routeClass);
+        if (routeAnnotation1 != null && routeAnnotation2 == null) {
+            return -1;
+        }
+        if (routeAnnotation1 == null && routeAnnotation2 != null) {
+            return 1;
+        }
+        return 0;
```
Comment (mcollovati) [inline flow-server/src/main/java/com/vaadin/flow/server/frontend/scanner/FrontendDependencies.java:892]: I wonder if we should compare not annotated classes by hierarchy depth, to prevent issues with deep hierarchy
For example:

```
@Route
ParentRoute {}

ChildRoute < ParentRoute

GrandchildRoute < ChildRoute
```

Comment (mshabarov) [reply flow-server/src/main/java/com/vaadin/flow/server/frontend/scanner/FrontendDependencies.java:892]: Good idea, one may use this kind of hierarchy. I'll extend the implementation to take it into account.

Comment (mshabarov) [reply flow-server/src/main/java/com/vaadin/flow/server/frontend/scanner/FrontendDependencies.java:892]: This might be enough to handle this case:
```java
if (routeAnnotation1 != null && routeAnnotation2 == null) {
            return -1;
        }
        if (routeAnnotation1 == null && routeAnnotation2 != null) {
            return 1;
        }
        if (entryPoint1.isAssignableFrom(entryPoint2)) {
            return -1;
        }
        if (entryPoint2.isAssignableFrom(entryPoint1)) {
            return 1;
        }
        return 0;
```

Comment (mshabarov) [reply flow-server/src/main/java/com/vaadin/flow/server/frontend/scanner/FrontendDependencies.java:892]: Don't know if we need to check hierarchy for annotated classes. Probably not, because they get proper reachable classes always.

Comment (mcollovati) [reply flow-server/src/main/java/com/vaadin/flow/server/frontend/scanner/FrontendDependencies.java:892]: The check for annotated classes is probably not needed, but we can maybe add a test to be sure it works correctly.

```
@Route("parent")
ParentRoute {}

@Route("child")
ChildRoute < ParentRoute
```

Comment (mshabarov) [reply flow-server/src/main/java/com/vaadin/flow/server/frontend/scanner/FrontendDependencies.java:913]: Done. Added few more tests to check the edge cases (ordered and reversed) and also case for shuffled entry points.

File: flow-server/src/test/java/com/vaadin/flow/server/frontend/scanner/FrontendDependenciesTest.java (hunk 2 above)
Comment (mshabarov) [reply flow-server/src/test/java/com/vaadin/flow/server/frontend/scanner/FrontendDependenciesTest.java]: Added more tests. JUnit 4 doesn't have "per method" parameterised feature, so I made it with just three separate methods.

**Review by mcollovati — APPROVED** [review]
